File: ai-solutions/02/opus-4.5/rust/src/main.rs
use std::env;
use std::fs;

fn parse_input(filename: &str) -> Vec<(u64, u64)> {
    let content = fs::read_to_string(filename).expect("Failed to read input file");
    let mut ranges = Vec::new();
    
    for part in content.trim().split(',') {
        let trimmed = part.trim();
        if !trimmed.is_empty() {
            let parts: Vec<&str> = trimmed.split('-').collect();
            let start: u64 = parts[0].parse().expect("Invalid start number");
            let end: u64 = parts[1].parse().expect("Invalid end number");
            ranges.push((start, end));
        }
    }
    ranges
}

fn is_double_repeated(n: u64) -> bool {
    // Part 1: Check if number is a pattern repeated exactly twice
    let s = n.to_string();
    let length = s.len();
    
    if length % 2 != 0 {
        return false;
    }
    
    let half = length / 2;
    s[..half] == s[half..]
}

fn is_repeated_pattern(n: u64) -> bool {
    // Part 2: Check if number is a pattern repeated at least twice
    let s = n.to_string();
    let length = s.len();
    
    for pattern_len in 1..=(length / 2) {
        if length % pattern_len == 0 {
            let repetitions = length / pattern_len;
            if repetitions >= 2 {
                let pattern = &s[..pattern_len];
                let repeated: String = pattern.repeat(repetitions);
                if repeated == s {
                    return true;
                }
            }
        }
    }
    false
}

fn solve(ranges: &[(u64, u64)]) -> (u64, u64) {
    let mut part1_sum: u64 = 0;
    let mut part2_sum: u64 = 0;
    
    for &(start, end) in ranges {
        for num in start..=end {
            if is_double_repeated(num) {
                part1_sum += num;
            }
            if is_repeated_pattern(num) {
                part2_sum += num;
            }
        }
    }
    
    (part1_sum, part2_sum)
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let input_file = if args.len() > 1 {
        &args[1]
    } else {
        "inputs/02.txt"
    };
    let ranges = parse_input(input_file);
    let (part1, part2) = solve(&ranges);
    println!("{}", part1);
    println!("{}", part2);
}

File: ai-solutions/02/opus-4.5/python/main.py
import sys


def parse_input(filename):
    """Parse the input file and return a list of (start, end) tuples."""
    with open(filename, 'r') as f:
        content = f.read().strip()
    
    ranges = []
    for part in content.split(','):
        part = part.strip()
        if part:
            start, end = part.split('-')
            ranges.append((int(start), int(end)))
    return ranges


def is_double_repeated(n):
    """Part 1: Check if number is a pattern repeated exactly twice."""
    s = str(n)
    length = len(s)
    if length % 2 != 0:
        return False
    half = length // 2
    return s[:half] == s[half:]


def is_repeated_pattern(n):
    """Part 2: Check if number is a pattern repeated at least twice."""
    s = str(n)
    length = len(s)
    for pattern_len in range(1, length // 2 + 1):
        if length % pattern_len == 0:
            repetitions = length // pattern_len
            if repetitions >= 2:
                pattern = s[:pattern_len]
                if pattern * repetitions == s:
                    return True
    return False


def solve(ranges):
    """Find all invalid IDs and return sums for Part 1 and Part 2."""
    part1_sum = 0
    part2_sum = 0
    
    for start, end in ranges:
        for num in range(start, end + 1):
            if is_double_repeated(num):
                part1_sum += num
            if is_repeated_pattern(num):
                part2_sum += num
    
    return part1_sum, part2_sum


def main():
    input_file = sys.argv[1] if len(sys.argv) > 1 else 'inputs/02.txt'
    ranges = parse_input(input_file)
    part1, part2 = solve(ranges)
    print(part1)
    print(part2)


if __name__ == '__main__':
    main()
